sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/common/CounterName.java

```diff
@@ -0,0 +1,155 @@
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A structured counter name.
+ *
+ * <p>It holds all information for generating structured name protos.
+ * It is also used as a wrapper during the migration to structured counters.
+ *
+ * <p>Before migration, the {@link CounterName} will be converted to a flat name.
+ * During migration, the {@link CounterName} will be mapped to a structured name proto.
+ */
+public class CounterName {
+  /**
+   * Returns a {@link CounterName} with the given name.
+   */
+  public static CounterName named(String name) {
+    return new CounterName(name, "", "", "");
+  }
+
+  /**
+   * Returns a msecs {@link CounterName}.
+   */
+  public static CounterName msecs(String name) {
+    return named(name + "-msecs");
+  }
+
+  /**
+   * Returns a {@link CounterName} identical to this, but with the given origin.
+   */
+  public CounterName withOrigin(String origin) {
+    return new CounterName(this.name, origin, this.stepName, this.prefix);
+  }
+
+  /**
+   * Returns a {@link CounterName} identical to this, but with the given step name.
+   */
+  public CounterName withStepName(String stepName) {
+    return new CounterName(this.name, this.origin, stepName, this.prefix);
+  }
+
+  /**
+   * Returns a {@link CounterName} identical to this, but with the given prefix.
+   */
+  public CounterName withPrefix(String prefix) {
+    return new CounterName(this.name, this.origin, this.stepName, prefix);
+  }
+
+  /**
+   * Name of the counter.
+   *
+   * <p>For example, process-msecs, ElementCount.
+   */
+  private final String name;
+
+  /**
+   * Origin (namespace) of counter name.
+   *
+   * <p>For example, "user" for user-defined counters.
+   * It is empty for counters defined by the Dataflow service or SDK.
+   */
+  private final String origin;
+
+  /**
+   * System defined step name or the named-output of a step.
+   *
+   * <p>For example, {@code s1} or {@code s2.out}.
+   * It may be empty when counters don't associate with step names.
+   */
+  private final String stepName;
+
+  /**
+   * Prefix of group of counters.
+   *
+   * <p>It is empty when counters don't have general prefixes.
+   */
+  private final String prefix;
+
+  /**
+   * Flat name is the equivalent unstructured name.
+   *
+   * <p>It is null before {@link #getFlatName()} is called.
+   */
+  private AtomicReference<String> flatName;
+
+  private CounterName(String name, String origin, String stepName, String prefix) {
+    this.name = checkNotNull(name, "name");
+    this.origin = checkNotNull(origin, "origin");
+    this.stepName = checkNotNull(stepName, "stepName");
+    this.prefix = checkNotNull(prefix, "prefix");
+    this.flatName = new AtomicReference<String>();
+  }
+
+  /**
+   * Returns the flat name of a structured counter.
+   */
+  public String getFlatName() {
+    String ret = flatName.get();
+    if (ret == null) {
+      StringBuilder sb = new StringBuilder();
+      if (!Strings.isNullOrEmpty(prefix)) {
```

Comment:
Given that we've already checkNotNulled these fields, we should be able to blind-append them (and/or check empty directly)